protocols/randhound/proof.go

```diff
@@ -0,0 +1,298 @@
+// is computed as the hash over all base points and commitments.
+func (p *Proof) SetupCollective(scalar ...abstract.Scalar) ([]abstract.Point, []abstract.Point, []ProofCore, error) {
+
+	if len(scalar) != len(p.base) {
+		return nil, nil, nil, errors.New("Received unexpected number of scalars")
+	}
+
+	n := len(scalar)
+	p.core = make([]ProofCore, n)
+	v := make([]abstract.Scalar, n)
+	xG := make([]abstract.Point, n)
+	xH := make([]abstract.Point, n)
+	vG := make([]abstract.Point, n)
+	vH := make([]abstract.Point, n)
+	for i, x := range scalar {
+
+		xG[i] = p.suite.Point().Mul(p.base[i].g, x)
+		xH[i] = p.suite.Point().Mul(p.base[i].h, x)
+
+		// Commitments
+		v[i] = p.suite.Scalar().Pick(random.Stream)
+		vG[i] = p.suite.Point().Mul(p.base[i].g, v[i])
+		vH[i] = p.suite.Point().Mul(p.base[i].h, v[i])
+	}
+
+	// Collective challenge
+	cb, err := crypto.HashArgsSuite(p.suite, xG, xH, vG, vH)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	c := p.suite.Scalar().Pick(p.suite.Cipher(cb))
+
+	// Responses
+	for i, x := range scalar {
+		r := p.suite.Scalar()
+		r.Mul(x, c).Sub(v[i], r)
+		p.core[i] = ProofCore{c, r, vG[i], vH[i]}
+	}
+
+	return xG, xH, p.core, nil
+}
+
+// Verify validates the proof against the given input by checking that
+// vG == rG + c(xG) and vH == rH + c(xH).
+func (p *Proof) Verify(xG []abstract.Point, xH []abstract.Point) ([]int, []int, error) {
+
+	if len(xG) != len(xH) {
+		return nil, nil, errors.New("Received unexpected number of points")
+	}
+
+	var good, bad []int
+	for i := 0; i < len(p.base); i++ {
+		if xG[i].Equal(p.suite.Point().Null()) || xH[i].Equal(p.suite.Point().Null()) {
+			bad = append(bad, i)
+		} else {
+			rG := p.suite.Point().Mul(p.base[i].g, p.core[i].R)
+			rH := p.suite.Point().Mul(p.base[i].h, p.core[i].R)
+			cxG := p.suite.Point().Mul(xG[i], p.core[i].C)
+			cxH := p.suite.Point().Mul(xH[i], p.core[i].C)
+			a := p.suite.Point().Add(rG, cxG)
+			b := p.suite.Point().Add(rH, cxH)
+
+			if p.core[i].VG.Equal(a) && p.core[i].VH.Equal(b) {
+				good = append(good, i)
+			} else {
+				bad = append(bad, i)
+			}
+		}
+	}
+
+	return good, bad, nil
+}
+
+// PVSS implements public verifiable secret sharing.
+type PVSS struct {
+	suite abstract.Suite // Suite
+	h     abstract.Point // Base point for polynomial commits
+	t     int            // Secret sharing threshold
+}
+
+// NewPVSS creates a new PVSS struct using the given suite, base point, and
+// secret sharing threshold.
+func NewPVSS(s abstract.Suite, h abstract.Point, t int) *PVSS {
+	return &PVSS{suite: s, h: h, t: t}
+}
+
+// Split creates PVSS shares encrypted by the public keys in X and
+// provides a NIZK encryption consistency proof for each share.
+func (pv *PVSS) Split(X []abstract.Point, secret abstract.Scalar) ([]int, []abstract.Point, []ProofCore, []byte, error) {
+
+	n := len(X)
+
+	// Create secret sharing polynomial
+	p := new(poly.PriPoly).Pick(pv.suite, pv.t, secret, random.Stream)
```

Comment:
Possibility of renaming `p`, `s` and `P` to `poly`, `shares`, `commit`, so you don't need comments.